### DISCUSSION TITLE
Custom json encoder for datetime

### DISCUSSION THREAD

**Discussion by mquilling** [pydantic/pydantic#7199] [Pydantic V2]
I am using this in v2 to add custom serialization for ALL datetime fields in any model
```
class BaseModel(PydanticBaseModel):
    model_config = ConfigDict(json_encoders={
        datetime: _format_datetime
    })
```
json_encoders is deprecated - how can this be done in v2 without touching every datetime field?

**Comment by harunyasar** [comment]
You could use `@field_serializer` or `@model_serializer` decorator. For more information https://docs.pydantic.dev/latest/usage/serialization/#custom-serializers

**Reply by mquilling** [reply]
that requires it to be applied to each field/model individually, I just want to change the base model and not have to go add a decorator/function to every single datetime attribute/model (we have hundreds).

**Reply by harunyasar** [reply]
One thought that occurred to me is that you could potentially override an existing field serializer.

```python
def serialize_dt(self: Any, dt: datetime, _info: Any):
    return dt.timestamp()


class ParentModel(BaseModel):
    def __init__(self, *args, **kwargs):
        for field_name, field_info in self.model_fields.items():
            if field_info.annotation == datetime:
                setattr(self, f"serialize_dt_{field_name}",  field_serializer(field_name)(serialize_dt))
        
        super().__init__(*args, **kwargs)


class ChildModel(ParentModel):
    x: datetime
    y: datetime

    # serialize_dt_x = field_serializer("x")(serialize_dt)
    # serialize_dt_y = field_serializer("y")(serialize_dt)
```

Obviously, this is not going to work. Nevertheless, it's an idea illustrating a potential method. It would be valuable to ascertain if there exists a means to dynamically establish these `serialize_dt_*` class variables in runtime. It's plausible that someone more knowledgeable on the subject might exist.

**Reply by adriangb** [reply]
See https://docs.pydantic.dev/latest/usage/types/custom/#adding-validation-and-serialization

**Reply by janbuchar** [reply]
I may be mistaken, but doesn't this require rewriting every usage of datetime to the annotated type? That may prove tedious and also error prone if we want to enforce a certain datetime format throughout the whole app.

**Reply by adriangb** [reply]
Yes it does, but it's also more explicit. Pros and cons. We are considering alternatives.

**Reply by lambdaq** [reply]
@adriangb  any new plans? I am switching back to v1 for the moment. 

I need a unified serialization for my customized type across every model definitions.

**Reply by adriangb** [reply]
No we probably won’t implement what you’re asking for any time soon

**Reply by ghost** [reply]
It's a bummer you think that removing this has any pro's. A user should be able to (but not required to) set default serializers for any and all types, forcing explicit serialization rules creates duplicate code and is not a great pattern to copy/paste the same thing in dozens of models. 

The first thing I do when I encounter duplicate code is refactor, so this change seems like your going backwards and making a user write more code to achieve the same result. More code almost always equals higher chances of bugs and more difficult to maintain.

Sounds like this TODO will be here a while, lol.
```
from pydantic import BaseModel as PydanticBaseModel

class BaseModel(PydanticBaseModel):
    # TODO there is not a 1:1 replacement for this in pydantic v2 :(
    #   -W ignore::pydantic.warnings.PydanticDeprecatedSince20
    model_config = ConfigDict(json_encoders={
        datetime: _format_datetime
    })
```

**Reply by harunyasar** [reply]
> I may be mistaken, but doesn't this require rewriting every usage of datetime to the annotated type? That may prove tedious and also error prone if we want to enforce a certain datetime format throughout the whole app.

Why don't you want simply create your own data type? By doing this, you will need to only import your new data type to wherever you want to use it.

```python
from datetime import datetime
from typing_extensions import Annotated

from pydantic import PlainSerializer, BaseModel

CustomDateTime = Annotated[
    datetime,
    PlainSerializer(lambda _datetime: _datetime.strftime("%m/%d/%Y, %H:%M:%S"), return_type=str),
]

class ExampleModel(BaseModel):
    custom_datetime: CustomDateTime

m = ExampleModel(custom_datetime=datetime.now())

m.model_dump()  # {'custom_datetime': '12/08/2023, 10:42:31'}
```

**Reply by janbuchar** [reply]
1. migration requires touching each and every datetime field in every model
2. we would need to pay attention every time we add a new field not to add a datetime instead of the custom type - this is super easy to miss and the errors can be very subtle

It is true that a custom type is more explicit, but so is writing a separate serializer function for every model. Yet people still choose to abstract this.

**Reply by harunyasar** [reply]
@janbuchar fair enough

**Comment by janbuchar** [comment]
You can accomplish a lot by setting the model serializer for your base model class - here's an example of adjusting the model attributes before and after the default serialization takes place:

```python
from typing import Any, Callable, Self
from pydantic import BaseModel, model_serializer
from datetime import datetime, timedelta, timezone


class ParentModel(BaseModel):
    foo: datetime

    @model_serializer(mode="wrap")
    def serialize(
        self, original_serializer: Callable[[Self], dict[str, Any]]
    ) -> dict[str, Any]:
        for field_name, field_info in self.model_fields.items():
            if field_info.annotation == datetime:
                setattr(
                    self,
                    field_name,
                    getattr(self, field_name).replace(tzinfo=timezone.utc),
                )

        result = original_serializer(self)

        for field_name, field_info in self.model_fields.items():
            if field_info.annotation == timedelta:
                result[field_name] = getattr(self, field_name).total_seconds()

        return result


class ChildModel(ParentModel):
    bar: datetime
    baz: timedelta
    qux: list[datetime]


foo = ChildModel(
    foo=datetime(2023, 3, 3, 3, 3),
    bar=datetime(2024, 4, 4, 4, 4),
    baz=timedelta(minutes=30, microseconds=10),
    qux=[datetime(2023, 5, 5, 5, 5), datetime(2023, 6, 6, 6, 6)],
)
print(foo.model_dump_json())

```

**Reply by Alsiri0n** [reply]
Hello, at this moment I use deprecated json_encoders and my code looks like this. Maybe it can help solve this problem more elegant than iteration for all model fields.

```python
from datetime import datetime
from pydantic import create_model, ConfigDict
from typing import Union

def main():
    data_for_create_model = [
        {"name": "field_a", "type": int},
        {"name": "field_b", "type": str},
        {"name": "field_c", "type": bool},
        {"name": "field_d", "type": datetime},
    ]
    temp_fields = {}
    for el in data_for_create_model:
        temp_fields.setdefault(f'{el["name"]}', (Union[el["type"], None], None))

    temp_model = create_model(
        "temp_model",
        **temp_fields,
        __config__=ConfigDict(json_encoders={datetime: datetime_serializers}),
    )

    data_for_validate_1 = {
        "field_a": 1,
        "field_b": "1",
        "field_c": True,
        "field_d": "2023-10-27 17:42:42",
    }
    pydantic_model_1 = temp_model(**data_for_validate_1)
    print(pydantic_model_1.model_dump_json())

def datetime_serializers(dt: datetime) -> str:
    pattern = "%Y-%m-%dT%H:%M:%S.%fZ"
    return dt.strftime(pattern)
```
```bash
>>> {"field_a":1,"field_b":"1","field_c":true,"field_d":"2023-10-27T17:42:42.000000Z"}
```

**Reply by skuechlin** [reply]
@janbuchar will your solution work for Optional[datetime] as well, or do we have to detect all cases of nested datetime List[datetime], etc. explicitly?

**Reply by janbuchar** [reply]
@skuechlin I updated the example for you, it works for a `list[datetime]`. I believe that pydantic is able to pick apart types such as list or optional.

**Reply by skuechlin** [reply]
thanks, @janbuchar . When I run the example, it gives me
{"foo":"2023-03-03T03:03:00Z","bar":"2024-04-04T04:04:00Z","baz":1800.00001,"qux":["2023-05-05T05:05:00","2023-06-06T06:06:00"]}
so the datetimes in the list are formatted differently. Might be out of scope for OP's question, but I was wondering if there was a generic solution in v2 that could control all datetime formatting on model output, regardless of nesting (as long as the annotation is correct, of course...

**Reply by adriangb** [reply]
https://github.com/pydantic/pydantic/discussions/7199#discussioncomment-6841946

**Reply by janbuchar** [reply]
@skuechlin yeah, you're right. So you'd probably have to analyze the types yourself and adjust the values accordingly. I doubt you can use any helper from pydantic, that'll probably be written in Rust somewhere in `pydantic-core`.

**Comment by wooparadog** [comment]
We're migrating to V2 and encountered this problem as well, however the use case is a little different:

```python
from pydantic import *
import datetime


class A(BaseModel):
    start_at: datetime.datetime


class B(BaseModel):
    a: A
    model_config = ConfigDict(json_encoders={datetime.datetime: lambda v: v.strftime('%Y-%m-%d %H:%M')})


print(
    B(a=A(start_at=datetime.datetime.now())).model_dump_json()
)
```


And in v1 A.start_at would also be formatted using config in B. it's not working in V2 though

**Comment by bverhoeve** [comment]
We also encountered this problem. In our case we have 2 models, one of which is used a more basic type (`TimeStart` across the application and a second one that uses that (`Appointment`):
```python
class TimeStart(BaseModel):
    time_start: time

class Appointment(BaseModel):
    appointment_time: TimeStart

    class Config:
        json_encoders = {
            time: lambda v: v.strftime("%H:%M"),
        }
```
The code above is simplified to illustrate the issue. An `Appointment` should serialise without seconds granularity as this is often not needed for appointments between humans. But `TimeStart` can also be used for other time intervals where seconds granularity is important. By using the `json_encoders` like in V1 on `Appointment` this can be achieved.

The suggested approach in V2 would be to use an annotated type in `TimeStart`, but that would make the default use case for `TimeStart` to serialize without seconds granularity. Another option would be to add it in `model_serializer` or `field_serializer` of the `Appointment` model to somehow inject it when serializing the nested `TimeStart` model. As soon as your models become complex enough, this becomes very cumbersome to write and maintain. Having a serialization option on the top-level model that is applied automatically to each type no matter its place in the hierarchy was a clean solution to that...

**Comment by ghostbody** [comment]
We need global encoder behavior to dump all `datetime` fields(for every model) to timestamps.

We achieve this using a ugly hack in v1:

```python
import pydantic.json
pydantic.json.ENCODERS_BY_TYPE[datetime] = lambda dt: int(dt.timestamp())
```

But, it's definitely not work in v2. 😢 

I don't know how to deal with this, as there is handreds of `Model` class which inhreit from `Pydantic.BasModel` in my project.
